src/modules/dotnet.rs

```diff
@@ -0,0 +1,308 @@
+use ansi_term::Color;
+use std::ffi::OsStr;
+use std::iter::Iterator;
+use std::ops::Deref;
+use std::path::Path;
+use std::process::{Command, Stdio};
+use std::str;
+
+use super::{Context, Module};
+
+type JValue = serde_json::Value;
+
+const DOTNET_SYMBOL: &str = "•NET ";
+const GLOBAL_JSON_FILE: &str = "global.json";
+const PROJECT_JSON_FILE: &str = "project.json";
+
+/// A module which shows the latest (or pinned) version of the dotnet SDK
+///
+/// Will display if any of the following files are present in
+/// the current directory:
+/// global.json, project.json, *.sln, *.csproj, *.fsproj, *.xproj
+pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
+    let dotnet_files = get_local_dotnet_files(context).ok()?;
+    if dotnet_files.is_empty() {
+        return None;
+    }
+
+    let mut module = context.new_module("dotnet");
+
+    // Internally, this module uses its own mechanism for version detection.
+    // Typically it is twice as fast as running `dotnet --version`.
+    let enable_heuristic = module.config_value_bool("heuristic").unwrap_or(true);
+    let version = if enable_heuristic {
+        let repo_root = context
+            .get_repo()
+            .ok()
+            .and_then(|r| r.root.as_ref().map(|buf| buf.as_path()));
+        estimate_dotnet_version(&dotnet_files, &context.current_dir, repo_root)?
+    } else {
+        get_version_from_cli()?
+    };
+
+    let module_style = module
+        .config_value_style("style")
+        .unwrap_or_else(|| Color::Blue.bold());
+
+    module.set_style(module_style);
+    module.new_segment("symbol", DOTNET_SYMBOL);
+    module.new_segment("version", &version);
+
+    Some(module)
+}
+
+fn estimate_dotnet_version<'a>(
+    files: &[DotNetFile<'a>],
+    current_dir: &Path,
+    repo_root: Option<&Path>,
+) -> Option<Version> {
+    let get_file_of_type = |t: FileType| files.iter().find(|f| f.file_type == t);
+
+    // Check for a global.json or a solution file first, but otherwise take the first file
```

Comment:
*Why* should we take the first file though?

Reply:
Ah, good point, this isn't entirely clear. In reality, it's more of a "take any" rather than "take first" situation. I'll update the comment.